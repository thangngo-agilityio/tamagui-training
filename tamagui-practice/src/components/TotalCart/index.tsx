'use client';

import { Button, Heading, HStack, Stack, Text, VStack } from '@chakra-ui/react';
import { useCallback, useEffect, useMemo, useState } from 'react';
import { useRouter } from 'next/navigation';

// Actions
import { updateMyCart } from '@/actions';

// hooks
import { useCustomToast } from '@/hooks';

// Constants
import { ROUTER, SUCCESS_MESSAGES } from '@/constants';

// Types
import { ICartItem } from '@/types';

// Utils
import { calculateTotalPrice, formatAmountNumber } from '@/utils';

type TTotalCart = {
  cartItem: ICartItem[];
  cartId: string;
};

const TotalCart = ({ cartItem, cartId }: TTotalCart) => {
  const [isDisable, setIsDisable] = useState(false);
  const { showToast } = useCustomToast();
  const router = useRouter();

  const total = useMemo(() => calculateTotalPrice(cartItem), [cartItem]);

  const handleCheckout = useCallback(async () => {
    const res = await updateMyCart(cartId, []);

    const { error } = res || {};

    if (error) {
      showToast(error);
    } else {
      showToast(SUCCESS_MESSAGES.CHECKOUT, 'success');
    }

    return router.push(ROUTER.ORDER);
  }, [cartId, showToast, router]);

  useEffect(() => {
    if (cartItem.length === 0) {
      setIsDisable(true);
    } else {
      setIsDisable(false);
    }
  }, [cartItem]);

  return (
    <Stack
      flexDir={{ base: 'column', lg: 'row' }}
      justifyContent="space-between"
    >
      <VStack alignItems="flex-start" mb={{ base: '20px', lg: 'unset' }}>
        <HStack gap="15px">
          <Heading
            variant="productTitle"
            size={{ base: 'size4xl', lg: 'size7xl' }}
          >
            Total:
          </Heading>
          <Text variant="totalCart" size={{ base: 'text4Xl', lg: 'text10xl' }}>
            N{formatAmountNumber(total.toString())}
          </Text>
        </HStack>
        <Text variant="septenary" size={{ base: 'text2Xl', lg: 'text4Xl' }}>
          Delivery exclusive
        </Text>
      </VStack>
      <Button
        variant="checkout"
        size={{ base: 'xl', lg: 'size4xlSecond' }}
        onClick={handleCheckout}
        isDisabled={isDisable}
      >
        Checkout
      </Button>
    </Stack>
  );
};

export default TotalCart;
