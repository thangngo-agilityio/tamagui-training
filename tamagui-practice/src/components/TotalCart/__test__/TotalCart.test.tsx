import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import TotalCart from '../index';
import '@testing-library/jest-dom';

// Mock the formatAmountNumber function
jest.mock('@/utils', () => ({
  formatAmountNumber: jest.fn((amount) => amount),
}));

describe('TotalCart() TotalCart method', () => {
  const total = 1000;
  const isDisable = false;
  const onClick = jest.fn();
  it('should match snapshot for TotalCart', () => {
    const { container } = render(
      <TotalCart total={total} isDisable={isDisable} onClick={onClick} />,
    );

    expect(container).toMatchSnapshot();
  });
  describe('Happy Path', () => {
    it('should render the total amount correctly', () => {
      // Arrange

      // Act
      render(
        <TotalCart total={total} isDisable={isDisable} onClick={onClick} />,
      );

      // Assert
      expect(screen.getByText('Total:')).toBeInTheDocument();
      expect(screen.getByText(`N${total}`)).toBeInTheDocument();
    });

    it('should render the checkout button enabled when isDisable is false', () => {
      // Arrange
      const total = 1000;
      const isDisable = false;
      const onClick = jest.fn();

      // Act
      render(
        <TotalCart total={total} isDisable={isDisable} onClick={onClick} />,
      );

      // Assert
      const button = screen.getByRole('button', { name: /checkout/i });
      expect(button).toBeEnabled();
    });

    it('should call onClick handler when checkout button is clicked', () => {
      // Arrange
      const total = 1000;
      const isDisable = false;
      const onClick = jest.fn();

      // Act
      render(
        <TotalCart total={total} isDisable={isDisable} onClick={onClick} />,
      );
      const button = screen.getByRole('button', { name: /checkout/i });
      fireEvent.click(button);

      // Assert
      expect(onClick).toHaveBeenCalledTimes(1);
    });
  });

  describe('Edge Cases', () => {
    it('should render the checkout button disabled when isDisable is true', () => {
      // Arrange
      const total = 1000;
      const isDisable = true;
      const onClick = jest.fn();

      // Act
      render(
        <TotalCart total={total} isDisable={isDisable} onClick={onClick} />,
      );

      // Assert
      const button = screen.getByRole('button', { name: /checkout/i });
      expect(button).toBeDisabled();
    });

    it('should handle zero total amount gracefully', () => {
      // Arrange
      const total = 0;
      const isDisable = false;
      const onClick = jest.fn();

      // Act
      render(
        <TotalCart total={total} isDisable={isDisable} onClick={onClick} />,
      );

      // Assert
      expect(screen.getByText('Total:')).toBeInTheDocument();
      expect(screen.getByText('N0')).toBeInTheDocument();
    });

    it('should handle negative total amount gracefully', () => {
      // Arrange
      const total = -100;
      const isDisable = false;
      const onClick = jest.fn();

      // Act
      render(
        <TotalCart total={total} isDisable={isDisable} onClick={onClick} />,
      );

      // Assert
      expect(screen.getByText('Total:')).toBeInTheDocument();
      expect(screen.getByText('N-100')).toBeInTheDocument();
    });
  });
});
