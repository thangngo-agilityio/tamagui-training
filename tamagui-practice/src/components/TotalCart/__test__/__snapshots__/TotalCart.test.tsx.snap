// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TotalCart() TotalCart method should match snapshot for TotalCart 1`] = `
<div>
  <div
    class="chakra-stack css-1b3w1ka"
  >
    <div
      class="chakra-stack css-1sq0ttz"
    >
      <div
        class="chakra-stack css-14xzl1v"
      >
        <h2
          class="chakra-heading css-0"
        >
          Total:
        </h2>
        <p
          class="chakra-text css-0"
        >
          N
          1000
        </p>
      </div>
      <p
        class="chakra-text css-0"
      >
        Delivery exclusive
      </p>
    </div>
    <button
      class="chakra-button css-4xx2wk"
      type="button"
    >
      Checkout
    </button>
  </div>
</div>
`;
